### PR TITLE
Augment wildcard_match to allow underscore in domain name

### PR DESCRIPTION
**Pull request by shinrich**
Dave Carlin noticed that our name-based server certificate verifications were failing when the SNI value contained an underscore.  

```
[Mar  8 21:06:48.070] {0x2b5575c26700} [ET_NET 4] WARNING: SNI (springbot_www_production.s3.amazonaws.com) not in certificate. Action=Terminate server=springbot_www_production.s3.amazonaws.com(52.216.85.211)
```

Tracked it down to a valid character check in wildcard_match function that only allowed alpha numerics and hyphen.  Reading RFC 1034/1035, those are the suggested character sets for domain names, but those are not required restrictions.  Since AWS frequently uses underscore in their S3 domain names, it seems like we need to make this change in practice in any case.

### CODE REVIEW

**Comment by shinrich** [comment]
Updated commit to add a test case to the tls_verify test to exercise this issue.  Without this fix, tr5 will fail.

File: src/tscore/X509HostnameValidator.cc [hunk 1]
```diff
@@ -166,7 +166,7 @@ wildcard_match(const unsigned char *prefix, size_t prefix_len, const unsigned ch
    * permitted characters and only matches a single label
    */
   for (p = wildcard_start; p != wildcard_end; ++p) {
-    if (!(('0' <= *p && *p <= '9') || ('A' <= *p && *p <= 'Z') || ('a' <= *p && *p <= 'z') || *p == '-')) {
+    if (!(('0' <= *p && *p <= '9') || ('A' <= *p && *p <= 'Z') || ('a' <= *p && *p <= 'z') || *p == '-' || *p == '_')) {
```
Comment (bryancall) [inline src/tscore/X509HostnameValidator.cc]: This should be ordered to have the most likely match first.  I assume it should be lowercase, digit, uppercase, and then symbols.

**Comment by shinrich** [comment]
pushed new commit that reorders the character tests.

**Comment by randall** [comment]
[approve ci clang-analyzer]

**Pull request merged**